### PR TITLE
Invert text search facet/text filter

### PR DESCRIPTION
**Pull request by ostephens**
Implementing an 'Invert' option for the text search facet (a.k.a filter) as per #952 

* New 'Invert' and 'Reset' options at the top left of the Text filter facet panel
* Click the 'Invert' link to invert the filter - i.e. find all rows which don't match the filter text or regular expression
* When 'Invert' is applied, panel gets an orange header to indicate the filter is inverted. The same device also implemented on text facets. In both cases the 'Invert' link appears struck-through (this latter device is the current behaviour for the Text Facet)
* 'Reset' clears all the options (invert, text in the filter box, case-sensitive, regular expression) of the filter

Comments on the UI welcome (@ettorerizza, @thadguidry, @wetneb)

I want to implement some tests for this functionality, but I can't work out how to best do this - @jackyq2015 @wetneb @thadguidry - any hints on the right approach to writing a test for these? I'm not quite sure how to apply the filter with options to a test set of data to see what the results are (I know what I expect the results to be - I'm just not sure on the mechanics to get the results). Pointers v welcome.

### CODE REVIEW

**Comment by wetneb** [comment]
@ostephens I would write a test like this:
* create a sample project with a CSV importer
* add a facet that uses your inversion (this is done via the "Engine")
* create a RowVisitor and run it on the engine - the visitor will only be called for the rows that match the facet - you can do assertions in the visitor, or based on the visitor result.

File: main/src/com/google/refine/browsing/filters/ExpressionStringComparisonRowFilter.java [hunk 1]
```diff
@@ -67,44 +69,44 @@ public boolean filterRow(Project project, int rowIndex, Row row) {
...
-        
+        Boolean invert = _invert;
         Object value = _evaluable.evaluate(bindings);
         if (value != null) {
             if (value.getClass().isArray()) {
                 Object[] a = (Object[]) value;
                 for (Object v : a) {
                     if (checkValue(v instanceof String ? ((String) v) : v.toString())) {
-                        return true;
+                        return (invert) ? false : true;
```
Comment (wetneb) [inline main/src/com/google/refine/browsing/filters/ExpressionStringComparisonRowFilter.java]: I'd replace that by just `!invert`.

File: main/src/com/google/refine/browsing/filters/ExpressionStringComparisonRowFilter.java [hunk 2]
```diff
@@ -67,44 +69,44 @@ public boolean filterRow(Project project, int rowIndex, Row row) {
...
                         if (checkValue(a.get(i).toString())) {
-                            return true;
+                            return (invert) ? false : true;
                         }
                     } catch (JSONException e) {
                         // ignore
                     }
                 }
-                return false;
+                return (invert) ? true : false;
```
Comment (wetneb) [inline main/src/com/google/refine/browsing/filters/ExpressionStringComparisonRowFilter.java]: And that by `invert`.

**Comment by ostephens** [comment]
Thanks @wetneb - is there an example of a test already that uses a RowVisitor with assertions?

**Comment by wetneb** [comment]
@ostephens actually forget about that, I think there is a simpler way to test a facet: just run it on a row and check the result. So:
* instanciate your `Facet`
* use its `getRowFilter` to get a `RowFilter` out of it
* run it on a few rows from your project and check the result
Thanks for your effort, it's brave to write tests for a part of the code that is currently very poorly tested…  Once we have a few, it will be easier!

**Comment by jackyq2015** [comment]
@ostephens You can take a look of CacheTests.java. For the Javascript, I am not quite sure. There are some popular framework such as  [Karma](https://karma-runner.github.io) seems quite good. But I don't know how OR can fit into it.

**Comment by thadguidry** [comment]
Being test framework agnostic is a "good thing".
Owen, it probably would be wise to go headless with PhantomJS.   I like
that Karma can support it.

I myself have used Robot framework and also Sahi in the past.

+1 Karma seems to take a much easier route from my quick inspection.

-Thad
+ThadGuidry <https://plus.google.com/+ThadGuidry>

**Comment by ostephens** [comment]
Thanks for that hint @wetneb - now added tests

**Comment by ostephens** [comment]
@jackyq2015 @thadguidry not thinking of adding tests for the javascript/UI at the moment. I think there was an earlier attempt to add some tests for the JS using Windmill (https://github.com/windmill) but these have been disabled in the Travis build by #1227 because they were broken and Windmill discontinued

**Comment by ostephens** [comment]
Thanks all!

**Pull request merged**